### DISCUSSION TITLE
Berry file serving

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#19490] [Ideas]
Looking through the Berry webserver code, and the webclient code, I think there is room for a webserver function 'content_send_file' in webserver, similar to 'write_file' in webclient.
In theory, this could open the file, and connect the file as a stream to be sent on the connection (so in theory async?).

If Berry webserver.on() could have wildcards, then this would enable a simple generic webserver.  If we can't do wildcards (I have a feeling this is not easy), then at least it could serve files based on query string, or specific static locations.

Of course we CAN read the file and send it... but that does not enable the (possibly future) use of Stream.

This thought comes from the fact I'm using static file serving for images, but also thoughts about serving specific custom pages for control of specific features - e.g. a specific webcam page with many of the webcam commands featured.

thoughts?

**Comment by sfromis** [comment]
Are you aware that you already have a static file server as `http://tasmota32device/ufsd?download=/test.txt` ?

**Reply by btsimonh** [reply]
yep, but it's fixed mime type, etc.

**Comment by s-hadinger** [comment]
The current implementation is based on Arduino framework and there are 2 main problems:
- Arduino framework is very inflexible when it comes to patterns. There are no wildcards but prefixes, they catch everything and you can't remove them once you've set them. So it's difficult to get a good cooperation with other modules
- Arduino framework does not support async IO. The file must be downloaded at once, and if the client is slow at downloading the file, it freezes the entire Tasmota.

A good implementation with flexible wildcards and async IO is possible (I did async web client in Berry) but is a lot of work to reimplement a decent web server. I don't have time for this, but would welcome it and help, because there are good use cases for async web server.

**Reply by btsimonh** [reply]
that was my impression when I looked some time back.  The arduino stream stuff allows for memory to be minimised, but it's still effectively synchronous.

**Comment by Jason2866** [comment]
There is an Arduino Async Webserver https://github.com/me-no-dev/ESPAsyncWebServer
 Maybe could be used as boilerplate for a Berry implementation ?
Some projects do use as default in there Arduino Code. Most of them have issues with, since the code is written and optimized for esp8266.
Afaik there is no rock solid Async Web Server for ESP32x

**Reply by s-hadinger** [reply]
It's old code and I remember some instability with IR code reported by other projects.

It could be a good inspiration but someone needs to implement it
